File: Study/Programming Skills/03. Find The Index of The First Occurence in A String/JavaScript/main.js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function (haystack, needle) {
  switch (true) {
    case needle.length === 0:
      return 0;
    case haystack.length < needle.length:
      return -1;
    default:
      return haystack.indexOf(needle);
  }
};


// Test Case 1
const haystack = "hello", needle = "ll";
console.log(strStr(haystack, needle)); // 2

// Test Case 2
const haystack2 = "aaaaa", needle2 = "bba";
console.log(strStr(haystack2, needle2)); // -1

File: Study/Programming Skills/03. Find The Index of The First Occurence in A String/Go/main.go
package main

import "fmt"

func main() {
  fmt.Println(strStr("hello", "ll"))
  fmt.Println(strStr("aaaaa", "bba"))
}

func strStr(haystack string, needle string) int {
    if len(needle) == 0 {
        return 0
    }
    if len(haystack) < len(needle) {
        return -1
    }
    for i := 0; i < len(haystack) - len(needle) + 1; i++ {  
      if haystack[i] == needle[0] {
          if haystack[i:i+len(needle)] == needle {
              return i
          } // if the substring of the haystack is equal to the needle
      } // if the first character of the needle is found in the haystack
    } // for loop to iterate over the haystack
    return -1
  }

